_posts/2024-08-19-quarkus-metaprogramming.adoc

```diff
@@ -0,0 +1,372 @@
+    public PersonSerializer() {
+        super(Person.class);
+    }
+
+    public void serialize(Object obj, JsonGenerator jsonGen, SerializerProvider serProv) throws IOException {
+        Person person = (Person)obj;
+        jsonGen.writeStartObject();
+        jsonGen.writeNumberField("age", person.getAge());
+        jsonGen.writeStringField("name", person.getName());
+        jsonGen.writeEndObject();
+   }
+}
+----
+
+and then to add this serializer to a module and register it on the ObjectMapper used by Jackson to perform the json serialization.
+
+[source, java]
+----
+SimpleModule module = new SimpleModule();
+module.addSerializer(Person.class, new PersonSerializer());
+ObjectMapper mapper = new ObjectMapper();
+mapper.registerModule(module);
+----
+
+With this setup everytime Jackson will have to convert an instance of the Person class into its json representation it will do this through this custom serializer instead of accessing the person’s fields via reflection.
+
+The strategy to implement reflection-free json serialization is then quite straightforward, but of course it would be great if the Quarkus extension could generate and register those serializers automatically and effortlessly. The first step to achieve this is discovering at deployment time for which classes we need to generate those serializers, that in our case are the classes returned by all the REST endpoints in our application.
+
+== Discovering and indexing the classes to be serialized with Jandex
+
+In general https://quarkus.io/guides/cdi-reference#bean_discovery[bean discovery] is performed by Quarkus, which indexes all the beans that are part of the CDI process through https://smallrye.io/jandex/[Jandex]. Since it finds all the bean classes having a https://jakarta.ee/specifications/cdi/4.1/jakarta-cdi-spec-4.1.html#bean_defining_annotations[bean defining annotation], this already includes all services exposing one or more REST endpoints that must have such an annotation. This means that all endpoints are already discovered by default. Even better, the https://jakarta.ee/specifications/cdi/4.1/jakarta-cdi-spec-4.1.html#bean_defining_annotations[quarkus-rest] extension already collects https://github.com/quarkusio/quarkus/blob/4ee036a19d64f537bea7836a25d5362091b0e8fc/extensions/resteasy-reactive/rest/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveResourceMethodEntriesBuildItem.java#L30[an Entry for each rest method] in the `ResteasyReactiveResourceMethodEntriesBuildItem`.
+
+Here each of these entries contains an instance of the Jandex MethodInfo that carries all the necessary offline reflection information about the signature of a Java method implementing a specific REST endpoint. As anticipated we are interested in the types returned by those methods, because those are the classes of the objects that will require to be converted in json as response to the REST endpoints invocation.
```

Comment:
```suggestion
Here each of these entries contains an instance of the Jandex `MethodInfo` that carries all the necessary offline reflection information about the signature of a Java method implementing a specific REST endpoint. As anticipated we are interested in the types returned by those methods, because those are the classes of the objects that will require to be converted in json as response to the REST endpoints invocation.
```